File: projects/sdk-ts/test/tools.test.ts
import { strict as assert } from 'node:assert';
import { test } from 'node:test';
import {
  toolErrorCodes,
  ToolHandlerError,
  type ToolDeclaration,
  type ToolHandler,
  TOOLS_CALL_METHOD,
  TOOLS_DECLARE_METHOD,
} from '../src/tools.ts';

test('declaration roundtrips through JSON', () => {
  const d: ToolDeclaration = {
    name: 'stack.list',
    description: 'List stacks',
    input_schema: { type: 'object' },
    sensitivity: 'general',
  };
  const back = JSON.parse(JSON.stringify(d)) as ToolDeclaration;
  assert.equal(back.name, 'stack.list');
  assert.equal(back.sensitivity, 'general');
  assert.deepEqual(back.input_schema, { type: 'object' });
});

test('ToolHandlerError carries data', () => {
  const e = new ToolHandlerError('upstream rejected', { status: 500 });
  assert.equal(e.message, 'upstream rejected');
  assert.deepEqual(e.data, { status: 500 });
  assert.equal(e.name, 'ToolHandlerError');
  assert.ok(e instanceof Error);
});

test('handler signature accepts an async closure', async () => {
  const h: ToolHandler = async (args: unknown) => {
    const obj = args as { value: string };
    return { echoed: obj.value };
  };
  const out = (await h({ value: 'ping' })) as { echoed: string };
  assert.equal(out.echoed, 'ping');
});

test('method names and error codes match the locked contract', () => {
  assert.equal(TOOLS_DECLARE_METHOD, 'orca/tools.declare');
  assert.equal(TOOLS_CALL_METHOD, 'orca/tools.call');
  assert.equal(toolErrorCodes.UNKNOWN_TOOL, -32001);
  assert.equal(toolErrorCodes.SCHEMA_VIOLATION, -32002);
  assert.equal(toolErrorCodes.HANDLER_ERROR, -32003);
});

File: projects/sdk/kotlin/src/test/kotlin/orca/sdk/ToolsTest.kt
package orca.sdk

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

class ToolsTest {
    private val json = Json { ignoreUnknownKeys = true }

    @Test
    fun declarationRoundtrips() {
        val d = ToolDeclaration(
            name = "stack.list",
            description = "List stacks",
            input_schema = Json.parseToJsonElement("""{"type":"object"}"""),
            sensitivity = "general",
        )
        val s = json.encodeToString(ToolDeclaration.serializer(), d)
        val back = json.decodeFromString(ToolDeclaration.serializer(), s)
        assertEquals("stack.list", back.name)
        assertEquals("general", back.sensitivity)
    }

    @Test
    fun toolHandlerErrorCarriesData() {
        val data = buildJsonObject { put("status", 500) }
        val e = ToolHandlerError("upstream rejected", data)
        assertEquals("upstream rejected", e.message)
        assertNotNull(e.data)
        assertEquals(500, e.data!!.jsonObject["status"]!!.jsonPrimitive.content.toInt())
    }

    @Test
    fun handlerSamAcceptsLambda() = runBlocking {
        val h: ToolHandler = ToolHandler { args ->
            val v = args.jsonObject["value"]!!.jsonPrimitive.content
            buildJsonObject { put("echoed", v) }
        }
        val out = h.call(buildJsonObject { put("value", "ping") })
        assertEquals("ping", out.jsonObject["echoed"]!!.jsonPrimitive.content)
    }

    @Test
    fun protocolConstantsMatchTheLockedContract() {
        assertEquals("orca/tools.declare", ToolsProtocol.DECLARE_METHOD)
        assertEquals("orca/tools.call", ToolsProtocol.CALL_METHOD)
        assertEquals(-32001, ToolErrorCodes.UNKNOWN_TOOL)
        assertEquals(-32002, ToolErrorCodes.SCHEMA_VIOLATION)
        assertEquals(-32003, ToolErrorCodes.HANDLER_ERROR)
    }

    @Test
    fun resultPayloadShape() {
        val r = ToolCallResult(buildJsonObject { put("echoed", "ping") })
        val s = json.encodeToString(ToolCallResult.serializer(), r)
        assertTrue(s.contains("\"result\""))
        assertTrue(s.contains("\"echoed\""))
    }
}
